inception-pdf-editor/src/main/java/de/tudarmstadt/ukp/inception/pdfeditor/pdfanno/PdfAnnoRenderer.java

```diff
@@ -108,35 +110,62 @@ public static PdfAnnoModel render(AnnotatorState aState, VDocument aVDoc, String
+        int docTextLen = aDocumentText.length();
+        int vSpanBegin = aDocumentText.length();
+        int vSpanEnd = -1;
+        // search for begin of the first range and end of the last range and use those
+        for (VRange range : aVspan.getRanges())
+        {
+            if (range.getBegin() < vSpanBegin) vSpanBegin = range.getBegin();
+            if (range.getEnd() > vSpanEnd) vSpanEnd = range.getEnd();
         }
+
+        // use an offset before and after string to find a unique text snippet for an annotation
+        // begin with 0 and increase by 1 until a unique text is found
+        // offset is limited to length of a quarter of the text
+        for (int offset = 0; offset < docTextLen / 4 + 1; offset++)
+        {
+            // subtract offset from vSpanBegin and add offset to vSpanEnd and stay in bounds
+            int offsetBegin = vSpanBegin <= offset ? 0 : vSpanBegin - offset;
+            int offsetEnd = vSpanEnd < docTextLen - offset ? vSpanEnd + offset : docTextLen - 1;
+
+            // get annotated text from aDocumentText with offset before and after it
+            // also remove all whitespaces because they do not exist in PDFExtract text
+            String annotatedText = aDocumentText.substring(offsetBegin, offsetEnd).replaceAll("\\s", "");
+            String offsetBeforeText = aDocumentText.substring(offsetBegin, vSpanBegin).replaceAll("\\s", "");
+            String offsetAfterText = aDocumentText.substring(vSpanEnd, offsetEnd).replaceAll("\\s", "");
+
+            List<Integer> indices =
+                getAllIndices(aPdfExtractFile.getStringContent(), annotatedText);
+            if (indices.size() == 0) {
+                // if there are no matches end search as adding more characters won't help anymore
+                break;
+            } else if (indices.size() == 1) {
+                int index = indices.get(0);
+                // get begin and end position of the original annotationText within PDFExtract text
+                int annotationBegin = index + offsetBeforeText.length();
+                int annotationEnd = index + annotatedText.length() - offsetAfterText.length() - 1;
+                // get according PDFExtract file lines for begin and end of annotation
+                PdfExtractLine firstLine = aPdfExtractFile.getStringPdfExtractLine(annotationBegin);
+                PdfExtractLine lastLine = aPdfExtractFile.getStringPdfExtractLine(annotationEnd);
+                return new Span(aVspan.getVid().getId(), firstLine.getPage(), aColor,
+                    annotatedText.substring(annotationBegin - index, annotationEnd - index + 1),
+                    firstLine.getPosition(), lastLine.getPosition());
+            }
+        }
+        LOG.error("Could not map existing annotation with id " + aVspan.getVid().toString());
+        return null;
+    }
+
+    private static List<Integer> getAllIndices(String searchIn, String searchFor)
```

Comment:
What does this method do?

Reply:
It searches for all starting indices of a String `searchFor` in the String `searchIn`.
Should I add Javadoc?

Reply:
Actually, you don't have to find all indexes, right? There are only three cases you currently handle:
* no match
* exactly one match
* more than one match (once you find the second match, the search can be aborted because it is already an error case)

You could for example change the method such that it throws an exception with a suitable error message in the no-match and more-than-one-match case.

Reply:
Adding JavaDoc would be good and renaming the method might also help (e.g. `findMatch`).

Reply:
Yes, you are right, I'll do it like that.